cert/aziot-certd/src/lib.rs

```diff
@@ -195,654 +242,356 @@ impl UpdateConfig for Api {
-                                            Error::Internal(InternalError::CreateCert(Box::new(
-                                                err,
-                                            )))
-                                        })?;
-                                let identity_private_key = api
-                                    .key_engine
-                                    .load_private_key(&identity_private_key)
-                                    .map_err(|err| {
-                                        Error::Internal(InternalError::CreateCert(Box::new(err)))
-                                    })?;
-
-                                let x509 = est::create_cert(
-                                    csr.to_owned(),
-                                    url,
-                                    auth_basic,
-                                    Some((&identity_cert, &identity_private_key)),
-                                    trusted_certs_x509,
-                                    api.proxy_uri.clone(),
+                    )
+                    .ok_or_else(||
+                        format!(
+                            "cert {:?} is configured to be issued by EST, but EST URL is not configured",
+                            id
+                        )
+                    )?;
+
+                let trusted_certs = default
+                    .map(|default|
+                        default.trusted_certs
+                            .iter()
+                            .try_fold(Vec::new(), |mut acc, cert| {
+                                let pem = get_cert_inner(
+                                        &api.homedir_path,
+                                        &api.preloaded_certs,
+                                        cert
+                                    )?
+                                    .ok_or_else(||
+                                        format!(
+                                            "cert_issuance.est.trusted_certs contains unreadable cert {:?}",
+                                            cert
+                                        )
+                                    )?;
+
+                                acc.extend(X509::stack_from_pem(&pem)?);
+                                Ok::<_, BoxedError>(acc)
+                            })
+                    )
+                    .transpose()?
+                    .unwrap_or_default();
+
+                // NOTE: Hints for borrow checker since closure uses mutable reference.
+                let homedir_path = &api.homedir_path;
+                let preloaded_certs = &api.preloaded_certs;
+                let key_client = &api.key_client;
+                let key_engine = &mut api.key_engine;
+                let id_opt = auth.x509.as_ref()
+                    .map(|x509| &x509.identity)
+                    .and_then(move |CertificateWithPrivateKey { cert, pk }| {
+                        let cert = get_cert_inner(
```

Comment:
```suggestion
                let id_opt = auth.x509.as_ref()
                    .map(|x509| &x509.identity)
                    .and_then({
                        let homedir_path = &api.homedir_path;
                        let preloaded_certs = &api.preloaded_certs;
                        let key_client = &api.key_client;
                        let key_engine = &mut api.key_engine;
                        let id_opt = auth.x509.as_ref()
                        move |CertificateWithPrivateKey { cert, pk }| {
                            let cert = get_cert_inner(
```

This makes it clear that the bindings are for the sake of the closure, and also defines them in a smaller scope.

Eg https://github.com/Azure/iotedge/blob/cb33102e7ac5139bf8f016e95bcc93e511f1a86f/edgelet/edgelet-http/src/lib.rs#L270-L272